psi4/driver/procrouting/diis.py

```diff
@@ -181,33 +224,163 @@ def extrapolate(self, *args):
         if np.all(diagonals > 0):
             diagonals = diagonals ** (- 0.5)
             B = np.einsum("i,ij,j -> ij", diagonals, B, diagonals)
-            coeffs = np.linalg.lstsq(B, rhs, rcond=None)[0][:-1] * diagonals[:-1]
+            return np.linalg.lstsq(B, rhs, rcond=None)[0][:-1] * diagonals[:-1]
         else:
-            coeffs = np.linalg.lstsq(B, rhs, rcond=None)[0][:-1]
+            return np.linalg.lstsq(B, rhs, rcond=None)[0][:-1]
+
+    def adiis_energy(self, x):
+        x = transform_input(x)
+        return np.dot(self.adiis_linear, x) + np.einsum("i,ij,j->", x, self.adiis_quadratic, x) / 2
+
+    def adiis_gradient(self, x):
+        c = transform_input(x)
+        dedc = self.adiis_linear + np.einsum("i,ij->j", c, self.adiis_quadratic)
+
+        norm_sq = (x**2).sum()
+        dcdx = np.diag(x) * norm_sq - np.einsum("i,j->ij", x ** 2, x)
+        dcdx *= 2 / norm_sq**2
+
+        return np.einsum("i,ij->j", dedc, dcdx)
+
+    def adiis_coefficients(self):
+        self.adiis_populate()
+        result = minimize(self.adiis_energy, np.ones(len(self.stored_vectors)), method="BFGS",
+                jac = self.adiis_gradient, tol=1e-6, options={"maxiter": 200})
```

Comment:
Yes, I've checked this extensively. SciPy's BFGS may fail to converge in difficult cases if you push the convergence threshold lower, and there's no significant gain to tightening it in the cases where you can. (PySCF can use a lower threshold, because they wait an iteration before starting any DIIS variant.)